src/snowflake/snowpark/_internal/data_source/drivers/pymsql_driver.py

```diff
@@ -0,0 +1,234 @@
+    PymysqlTypeCode.BIGINT: IntegerType,
+    PymysqlTypeCode.YEAR: IntegerType,
+    PymysqlTypeCode.FLOAT: FloatType,
+    PymysqlTypeCode.DOUBLE: FloatType,
+    PymysqlTypeCode.CHAR: StringType,
+    PymysqlTypeCode.VARCHAR: StringType,
+    PymysqlTypeCode.TINYTEXT: StringType,
+    PymysqlTypeCode.TEXT: StringType,
+    PymysqlTypeCode.MEDIUMTEXT: StringType,
+    PymysqlTypeCode.LONGTEXT: StringType,
+    PymysqlTypeCode.ENUM: StringType,
+    PymysqlTypeCode.SET: StringType,
+    PymysqlTypeCode.BIT: StringType,
+    PymysqlTypeCode.BINARY: BinaryType,
+    PymysqlTypeCode.VARBINARY: BinaryType,
+    PymysqlTypeCode.TINYBLOB: BinaryType,
+    PymysqlTypeCode.BLOB: BinaryType,
+    PymysqlTypeCode.MEDIUMBLOB: BinaryType,
+    PymysqlTypeCode.LONGBLOB: BinaryType,
+    PymysqlTypeCode.DATE: DateType,
+    PymysqlTypeCode.DATETIME: TimestampType,
+    PymysqlTypeCode.TIMESTAMP: TimestampType,
+    PymysqlTypeCode.TIME: TimeType,
+    PymysqlTypeCode.JSON: VariantType,
+}
+
+
+class PymysqlDriver(BaseDriver):
+    def __init__(
+        self, create_connection: Callable[[], "Connection"], dbms_type: Enum
+    ) -> None:
+        super().__init__(create_connection, dbms_type)
+
+    def infer_schema_from_description(
+        self, table_or_query: str, cursor: "Cursor"
+    ) -> StructType:
+        if table_or_query.lower().startswith("select"):
+            cursor.execute(f"select A.* from ({table_or_query}) A limit 1")
+        else:
+            cursor.execute(f"select * from `{table_or_query}` limit 1")
+        data = cursor.fetchall()
+        raw_schema = cursor.description
+        raw_types = self.infer_type_from_data(data, len(raw_schema))
+
+        processed_raw_schema = []
+        for type, col in zip(raw_types, raw_schema):
+            new_col = list(col)
+            new_col[1] = PymysqlTypeCode((new_col[1], type))
+            processed_raw_schema.append(new_col)
+
+        self.raw_schema = processed_raw_schema
+        return self.to_snow_type(processed_raw_schema)
+
+    def to_snow_type(self, schema: List[Any]) -> StructType:
+        """
+        pymysql mysql type to type code mapping:
+        https://github.com/PyMySQL/PyMySQL/blob/main/pymysql/constants/FIELD_TYPE.py
+
+        mysql type to snowflake type mapping:
+        https://other-docs.snowflake.com/en/connectors/mysql6/view-data#mysql-to-snowflake-data-type-mapping
+        """
+        fields = []
+        for col in schema:
+            (
+                name,
+                type_code,
+                display_size,
+                internal_size,
+                precision,
+                scale,
+                null_ok,
+            ) = col
+            snow_type = BASE_PYMYSQL_TYPE_TO_SNOW_TYPE.get(type_code, None)
+            if snow_type is None:
+                raise NotImplementedError(f"mysql type not supported: {type_code}")
+            if type_code in (PymysqlTypeCode.DECIMAL, PymysqlTypeCode.NEWDECIMAL):
+                precision -= 2
```

Comment:
1. let's add some comment on why we did -2 here
2. let's call out the limitation as connector that we don't support precision beyond 38